### PR TITLE
The binary check handles unicode source files.

### PR DESCRIPTION
**Pull request by jish**
The `binary?` query makes a best guess as to whether a file is a binary
file or not. The previous check was reading some bytes form the
beginning of the file. It would then check to see what percentage of
those bytes were common ASCII characters. If the file was made up of
70% common ASCII characters, we would "guess" that the file was not
a binary file.

One of the failings of this test is that it does not take unicode
characters into account. To fix that, we attempt to convert all
unicode characters to the ASCII character `?` and then perform the
same calculation to see if 70% of the file contains common characters.

cc: @kyoshidajp @mpapis 
cf: https://github.com/jish/pre-commit/issues/238

### CODE REVIEW

**Comment by jish** [comment]
Ugh, dependencies have their own locked down dependencies...

https://travis-ci.org/jish/pre-commit/jobs/142644412

    Gem::InstallError: ruby_dep requires Ruby version >= 2.0.0, ~> 2.0.

https://travis-ci.org/jish/pre-commit/jobs/142644415

    Gem::InstallError: listen requires Ruby version >= 2.2.3, ~> 2.2.

File: lib/pre-commit/utils/staged_files.rb [hunk 1]
```diff
@@ -33,7 +33,10 @@ def staged_files_git_all
     private
       # from https://github.com/djberg96/ptools/blob/master/lib/ptools.rb#L90
       def binary?(file)
-        s = (File.read(file, File.stat(file).blksize) || "").split(//)
+        bytes = File.stat(file).blksize
+        bytes = 4096 if bytes > 4096
+        s = (File.read(file, bytes) || "")
+        s = s.encode('US-ASCII', :undef => :replace).split(//)
```
Comment (mpapis) [inline lib/pre-commit/utils/staged_files.rb]: I have a bad feeling about `encode` - I remember having some problems with it - could we try some iso/old format encoded file with this code?

**Comment by mpapis** [comment]
for the dependencies you can track the ruby version requirement change in the gems and then add as conditions in `Gemfile` - then bundler / rubygems should be able to handle it:
```ruby
if RUBY_VERSION <= "2.0.0"
  gem "listen", "..."
end
```
I remember having few of those on some gems

File: lib/pre-commit/utils/staged_files.rb (hunk 1 above)
Comment (jish) [inline lib/pre-commit/utils/staged_files.rb]: I'm just updating the code snippet used in `ptools` found here https://github.com/djberg96/ptools/blob/master/lib/ptools.rb#L90

I'm certainly open to using the approach that works best for the most people.

Comment (mpapis) [inline lib/pre-commit/utils/staged_files.rb]: ah if it's still rip off - I'm fine with it ;)

File: .travis.yml [hunk 2]
```diff
@@ -15,11 +15,10 @@ cache:
...
-  - ruby-1.9.3
-  - ruby-2.0.0
-  - ruby-2.1
-  - ruby-2.2
-  - ruby-2.3.0
+  - 2.0.0
+  - 2.1.9
+  - 2.2.5
+  - 2.3.1
+  - jruby-9.0.5.0
```
Comment (mpapis) [inline .travis.yml]: please move `jruby` on top as it takes longest - so the rest can finish when it's running

**Comment by mpapis** [comment]
LGTM - please move `jruby` and merge :)

**Comment by kyoshidajp** [comment]
@jish Thanks. I did some spot checks by this branch, then that files seems to be fine.

**Pull request merged**